cmd/traffic/cmd/agentinit/agent_init.go

```diff
@@ -8,77 +8,129 @@ import (
-		"-j", "REDIRECT", "--to-ports", agentPort)
-	if err != nil {
-		return fmt.Errorf("failed to append rule to %s: %w", inboundChain, err)
-	}
-	// Direct everything coming into PREROUTING into our own inbound chain.
-	// We do this as an append instead of an insert because this will prevent us from interfering with a service mesh
-	// if one exists. If a service mesh exists, its PREROUTING rules will kick in before ours, ensuring traffic
-	// coming into the pod does not bypass the mesh.
-	err = iptables.AppendUnique(nat, "PREROUTING",
-		"-p", cfg.AgentProtocol,
-		"-j", inboundChain)
-	if err != nil {
-		return fmt.Errorf("failed to append prerouting rule to direct to %s: %w", inboundChain, err)
-	}
-	// Any traffic heading out of the loopback and into the app port (other than traffic from the agent) needs to
-	// be redirected to the agent. This will ensure that if there's a service mesh, when the mesh's proxy goes to
-	// request the application, it will get a response via the traffic agent.
-	err = iptables.Insert(nat, "OUTPUT", 1, "-o", loopback,
-		"-m", "owner", "!", "--gid-owner", agentUID,
-		"-j", inboundChain)
-	if err != nil {
-		return fmt.Errorf("failed to insert ! --gid-owner rule in OUTPUT: %w", err)
+	defer cf.Close()
+
+	c := config{}
+	if err = yaml.NewDecoder(cf).Decode(&c.Config); err != nil {
+		return nil, fmt.Errorf("unable to decode agent ConfigMap: %w", err)
 	}
-	// Any agent traffic heading out on the loopback but NOT towards localhost needs to be processed in case
-	// it needs to be redirected. This is so that if the traffic agent requests its own IP, it doesn't just
-	// serve the app but actually goes through the agent, and thus through any intercepts.
-	// This is needed to support requesting an intercepted pod by IP (or to intercept a headless service).
-	err = iptables.Insert(nat, "OUTPUT", 1,
-		"-o", loopback,
-		"-p", cfg.AgentProtocol,
-		"!", "-d", "127.0.0.1/32",
-		"-m", "owner", "--gid-owner", agentUID,
-		"-j", inboundChain)
-	if err != nil {
-		return fmt.Errorf("failed to insert --gid-owner rule in OUTPUT: %w", err)
+	return &c, nil
+}
+
+func (c *config) configureIptables(ctx context.Context, iptables *iptables.IPTables, loopback string) error {
+	// These iptables rules implement routing such that a packet directed to the appPort will hit the agentPort instead.
+	// If there's no mesh this is simply request -> agent -> app (or intercept)
+	// However, if there's a service mesh we want to make sure we don't bypass the mesh, so the traffic
+	// will flow request -> mesh -> agent -> app
+
+	// A service mesh will typically use an UID different from the one used by this process
+	agentUID := strconv.Itoa(os.Getuid())
+
+	outputInsertCount := 0
+	for _, proto := range []string{"tcp", "udp"} {
+		hasRule := false
+		for _, cn := range c.Containers {
+			for _, ic := range cn.Intercepts {
+				if strings.EqualFold(proto, ic.Protocol) {
+					hasRule = true
+					break
```

Comment:
should this break the Containers loop instead of the Intercepts loop?

Reply:
Yes, it should.